iree/hal/cts/cts_test_base.h

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_CTS_CTS_TEST_BASE_H_
+#define IREE_HAL_CTS_CTS_TEST_BASE_H_
+
+#include <map>
+#include <mutex>
+#include <set>
+
+#include "iree/base/status.h"
+#include "iree/hal/api.h"
+#include "iree/testing/gtest.h"
+#include "iree/testing/status_matchers.h"
+
+namespace iree {
+namespace hal {
+namespace cts {
+
+// Common setup for tests parameterized across all registered drivers.
+class CtsTestBase : public ::testing::TestWithParam<std::string> {
+ protected:
+  virtual void SetUp() {
+    const std::string& driver_name = GetParam();
+
+    // Get driver with the given name and create its default device.
+    // Skip drivers that are (gracefully) unavailable, fail if creation fails.
+    auto driver_or = GetDriver(driver_name);
+    if (IsUnavailable(driver_or.status())) {
+      IREE_LOG(WARNING) << "Skipping test as driver is unavailable: "
+                        << driver_or.status();
+      GTEST_SKIP();
+      return;
+    }
+    IREE_ASSERT_OK_AND_ASSIGN(driver_, std::move(driver_or));
+
+    iree_hal_device_t* device;
+    iree_status_t status = iree_hal_driver_create_default_device(
+        driver_, iree_allocator_system(), &device);
+    IREE_ASSERT_OK(status);
+    device_ = device;
+    iree_string_view_t device_id = iree_hal_device_id(device_);
+
+    device_allocator_ = iree_hal_device_allocator(device_);
+    iree_hal_allocator_retain(device_allocator_);
+  }
+
+  virtual void TearDown() {
+    if (device_allocator_) {
+      iree_hal_allocator_release(device_allocator_);
+      device_allocator_ = nullptr;
+    }
+    if (device_) {
+      iree_hal_device_release(device_);
+      device_ = nullptr;
+    }
+    if (driver_) {
+      iree_hal_driver_release(driver_);
+      driver_ = nullptr;
+    }
+  }
+
+  iree_hal_driver_t* driver_ = nullptr;
+  iree_hal_device_t* device_ = nullptr;
+  iree_hal_allocator_t* device_allocator_ = nullptr;
+
+ private:
+  // Gets a HAL driver with the provided name, if available.
+  static StatusOr<iree_hal_driver_t*> GetDriver(
```

Comment:
I want to eventually delete iree/base/status.h so we may want to use iree_status_t here (and iree_make_status instead of the builders, etc). One option is to move it to iree/testing/ and use it just for status_matchers, but not sure if that's useful or if we just make the status matchers support iree_status_t and then we have less code that is test-only (and this code gets more useful as a reference as it's just using the public API types).

Reply:
Ok, updated to just `iree_status_t`. Makes sense to limit our usage where possible. Much less code to maintain if we can trim some of the C++.